fetch/fetch.go

```diff
@@ -0,0 +1,526 @@
+	}
+	return m
+}
+
+// responseBatch is the response struct send for a requestBatch. the responseBatch ID must be the same
+// as stated in requestBatch even if not all data is present
+type responseBatch struct {
+	ID        types.Hash32
+	Responses []responseMessage
+}
+
+// Config is the configuration file of the Fetch component
+type Config struct {
+	BatchTimeout      int
+	MaxRetiresForPeer int
+	BatchSize         int
+	RequestTimeout    int
+}
+
+type peers interface {
+	GetPeers() []p2ppeers.Peer
+}
+
+// MessageNetwork is a network interface that allows fetch to communicate with other nodes with 'fetch servers'
+type MessageNetwork struct {
+	*server.MessageServer
+	peers
+	log.Log
+}
+
+// NewMessageNetwork creates a new instance of the fetch network server
+func NewMessageNetwork(requestTimeOut int, net service.Service, protocol string, log log.Log) *MessageNetwork {
+	return &MessageNetwork{
+		server.NewMsgServer(net.(server.Service), protocol, time.Duration(requestTimeOut)*time.Second, make(chan service.DirectMessage, p2pconf.Values.BufferSize), log),
+		p2ppeers.NewPeers(net, log.WithName("peers")),
+		log,
+	}
+}
+
+// GetRandomPeer returns a random peer from current peer list
+func (f MessageNetwork) GetRandomPeer() p2ppeers.Peer {
+	peers := f.peers.GetPeers()
+	if len(peers) == 0 {
+		f.Log.Panic("cannot send fetch - no peers found")
+	}
+	rand.Seed(time.Now().Unix()) // initialize global pseudo random generator
+	return peers[rand.Intn(len(peers))]
+}
+
+type network interface {
+	GetRandomPeer() p2ppeers.Peer
+	SendRequest(msgType server.MessageType, payload []byte, address p2pcrypto.PublicKey, resHandler func(msg []byte), failHandler func(err error)) error
+}
+
+// Fetch is the main struct that contains network peers and logic to batch and dispatch hash fetch Requests
+type Fetch struct {
+	cfg                  Config
+	log                  log.Log
+	dbs                  map[Hint]database.Database
+	activeRequests       map[types.Hash32][]request
+	activeBatches        map[types.Hash32]requestBatch
+	net                  network
+	requestReceiver      chan request
+	batchRequestReceiver chan []request
+	batchTimeout         *time.Ticker
+	stop                 chan struct{}
+	activeReqM           sync.RWMutex
+	activeBatchM         sync.RWMutex
+}
+
+// NewFetch creates a new FEtch struct
+func NewFetch(cfg Config, network service.Service, logger log.Log) *Fetch {
+
+	srv := NewMessageNetwork(cfg.RequestTimeout, network, fetchProtocol, logger)
+
+	f := &Fetch{
+		cfg:             cfg,
+		log:             logger,
+		dbs:             make(map[Hint]database.Database),
+		activeRequests:  make(map[types.Hash32][]request),
+		net:             srv,
+		requestReceiver: make(chan request),
+		batchTimeout:    time.NewTicker(time.Second * time.Duration(cfg.BatchTimeout)),
+		stop:            make(chan struct{}),
+		activeBatches:   make(map[types.Hash32]requestBatch),
+	}
+
+	srv.RegisterBytesMsgHandler(fetch, f.FetchRequestHandler)
```

Comment:
it uses a method of NewMessageNetwork, whereas the `Fetch` struct uses reduced functionality and RegisterBytesMsgHandler is not used at all by it... Do you think I should add this?

Reply:
`Fetcher` may not be using `RegisterBytesMsgHandler`, but the opposite applies - `RegisterBytesMsgHandler` is using `FetchRequestHandler`, which is defined as a method of `Fetcher` because it's using its dbs.

So if you prefer it to have a reduced functionality, `Fetcher` should be decoupled entirely from listening to peers' requests. Otherwise, listening should all be triggered inside `Start`, and should be documented.